universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
Rebased.

`shared_ptr` and `weak_ptr` are as thread safe as a raw pointer.  

Constructing, copying or destroying `shared_ptr` or `weak_ptr` guarantees that adjusting the counters are atomic operations.

Dereferencing and using the pointed to object is not thread-safe.

The pointers, `shared_ptr` and `weak_ptr` are thread-safe, but accessing the pointed to object without some other thread safety consideration is not.

Being passed any kind of pointer does not imply or enforce any thread safety.  If we want/expect objects to be thread safe, then they require additional protection beyond `shared_ptr`.

`const` is the correct way to indicate that an object should not be modified. This does not control or indicate anything about whether an object can/should be deleted.  Additionally, accessing a `const` object is not thread safe. 

The ownership that the type of pointer indicates is of resource management, in this case memory.  Here is another table:

Pointer type | modifiability | ownership (can I delete it)
---|---|---
 `unque_ptr<T>`              | I can        modify this `T`   |   I own/manage its resources (memory).
 `unque_ptr<const T>`    | I can  not modify this `T`   |   I own/manage its resources (memory).
 `shared_ptr<T>`             | I can        modify this `T`   |   I share ownership/management.
 `shared_ptr<const T>`   | I can not  modify this `T`   |   I share ownership/management.
 `T*`                                  | I can        modify this `T`   |   I do not own it.
 `const T*`                        | I can not  modify this `T`   |   I do not own it.

Regarding `BuildingManager` the pointed to building types are not thread safe and using `shared_ptr` will not make them thread safe.